histomicsui/web_client/vue/components/EditHeatmapOrGridData.vue

```diff
@@ -0,0 +1,229 @@
+import ColorPickerInput from './ColorPickerInput.vue';
+export default {
+    props: ['elementData'],
+    emits: ['submit', 'cancel'],
+    components: {
+        ColorPickerInput
+    },
+    data() {
+        return {
+            type: this.elementData.type,
+            interpretation: this.elementData.interpretation,
+            radius: this.elementData.radius,
+            normalizeRange: this.elementData.normalizeRange,
+            colorRange: _.clone(this.elementData.colorRange),
+            colorRangeData: [],
+            rangeValues: _.clone(this.elementData.rangeValues),
+            minColor: this.elementData.minColor,
+            maxColor: this.elementData.maxColor,
+            stepped: this.elementData.stepped,
+            scaleWithZoom: this.elementData.scaleWithZoom,
+            validationErrors: []
+        };
+    },
+    computed: {
+        headerMessage() {
+            if (this.type === 'heatmap') {
+                return 'Edit Heatmap Element';
+            }
+            return 'Edit Grid Data Element';
+        }
+    },
+    methods: {
+        getColorString(color) {
+            return tinycolor(color).toString();
+        },
+        addColor(index) {
+            const newEntry = {
+                r: 0,
+                g: 0,
+                b: 0,
+                a: 0,
+                colorString: 'rgba(0, 0, 0, 0)',
+                value: 0
+            }
+            this.colorRangeData.splice(index + 1, 0, newEntry);
+            this.updateRangeDataKeys();
+        },
+        removeColor(index) {
+            this.colorRangeData.splice(index, 1);
+            this.updateRangeDataKeys();
+        },
+        updateRangeDataKeys() {
+            _.forEach(this.colorRangeData, (entry, index) => {
+                entry.key = `${index}-${entry.colorString}`;
+            });
+        },
+        submitClicked() {
+            this.validationErrors = [];
+            this.tryValidateForm();
+            if (this.validationErrors.length === 0) {
+                this.notifySubmit();
+            }
+        },
+        tryValidateForm() {
+            if (this.type === 'heatmap') {
+                const radiusAsFloat = parseFloat(this.radius);
+                if (isNaN(radiusAsFloat) || radiusAsFloat <= 0) {
+                    this.validationErrors.push('Radius must be a positive number');
+                }
+            }
+            if (this.normalizeRange) {
+                // check values
+                const invalidRangeValues = this.rangeValues.filter(
+                    (value) => value < 0 || value > 1).map(
+                    (value, index) => index + 1
+                );
+                if (invalidRangeValues.length) {
+                    this.validationErrors.push(
+                        'Invalid value(s) in row(s) ' + invalidRangeValues.join(', ') + '. When "Normalize Range" is checked, values must be between 0 and 1'
+                    );
+                }
+            }
```

Comment:
Values need to also be distinct.  As part of the submission, we should sort the list of range/colors by range value.

Reply:
Added additional validation in 4be2b4f, added sorting in 42edea6